cranelift/codegen/src/machinst/mod.rs

```diff
@@ -299,13 +300,35 @@ pub struct MachCompileResult {
     pub bb_edges: Vec<(CodeOffset, CodeOffset)>,
 }
 
-impl MachCompileResult {
+impl CompiledCode {
     /// Get a `CodeInfo` describing section sizes from this compilation result.
     pub fn code_info(&self) -> CodeInfo {
         CodeInfo {
             total_size: self.buffer.total_size(),
         }
     }
+
+    /// Emit machine code directly into raw memory.
+    ///
+    /// Write all of the function's machine code to the memory at `mem`.
+    ///
+    /// The machine code is not relocated.
+    /// Instead, any relocations can be obtained from this struct's field `buffer`.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe since it does not perform bounds checking on the memory buffer,
+    /// and it can't guarantee that the `mem` pointer is valid.
+    ///
+    /// Returns information about the emitted code and data.
+    #[deny(unsafe_op_in_unsafe_fn)]
+    pub unsafe fn emit_to_memory(&self, mem: *mut u8) -> CodeInfo {
```

Comment:
This is the last bit of the refactor I had in mind I think: could we turn this `emit_to_memory` into just an accessor on `self.buffer.data()`, and let the embedder do the unsafe dump-the-data-into-a-raw-pointer if they desire?

The major thing we get out of that is to remove the one remaining `unsafe` API from `cranelift-codegen`; and it feels more idiomatic to me now that we have a type `CompiledCode` that wraps up the result to have an accessor providing access to a part of it, rather than a method that splats that content somewhere (noun rather than verb basically). Does that make sense?

Reply:
Nice, it even leads to more beautiful call sites which sometimes don't even have to use `unsafe` :+1: